cmd/cmd_run.go

```diff
@@ -238,38 +149,61 @@ func runHandler(opts *runCmdOptions, cmd *cobra.Command, _ []string) error {
 				Level: compress.LevelDefault,
 			}))
 
-		defer func() {
-			if err := app.ShutdownWithTimeout(shutdownTimeout); err != nil {
-				logger.ErrorContext(ctx, "Error during shutdown HTTP server", slogx.Error(err))
-				return
-			}
-			logger.InfoContext(ctx, "HTTP server stopped gracefully")
-		}()
-
 		// Health check
 		app.Get("/", func(c *fiber.Ctx) error {
 			return errors.WithStack(c.SendStatus(http.StatusOK))
 		})
 
-		// mount http handlers from each http-enabled module
-		for module, handler := range httpHandlers {
-			if err := handler.Mount(app); err != nil {
-				logger.PanicContext(ctx, "Something went wrong, can't mount HTTP handler", slogx.Error(err), slogx.String("module", module))
-			}
-			logger.InfoContext(ctx, "Mounted HTTP handler", slogx.String("module", module))
-		}
+		return app, nil
+	})
 
-		go func() {
-			// stop main process if API stopped
-			defer stop()
+	// Initialize worker context to separate worker's lifecycle from main process
+	ctxWorker, stopWorker := context.WithCancel(context.Background())
+	defer stopWorker()
+
+	// Run modules
+	{
+		modules := lo.Uniq(conf.EnableModules)
+		modules = lo.Map(modules, func(item string, _ int) string { return strings.TrimSpace(item) })
+		modules = lo.Filter(modules, func(item string, _ int) bool { return item != "" })
+		for _, module := range modules {
+			ctx := logger.WithContext(ctxWorker, slogx.String("module", module))
 
-			logger.InfoContext(ctx, "Started HTTP server", slog.Int("port", conf.HTTPServer.Port))
-			if err := app.Listen(fmt.Sprintf(":%d", conf.HTTPServer.Port)); err != nil {
-				logger.PanicContext(ctx, "Something went wrong, error during running HTTP server", slogx.Error(err))
+			indexer, err := do.InvokeNamed[indexer.IndexerWorker](injector, module)
+			if err != nil {
+				if errors.Is(err, do.ErrServiceNotFound) {
+					return errors.Errorf("Module %q is not supported", module)
+				}
+				return errors.Wrapf(err, "can't init module %q", module)
 			}
-		}()
+
+			// Run Indexer
+			if !conf.APIOnly {
+				go func() {
+					// stop main process if indexer stopped
+					defer stop()
+
+					logger.InfoContext(ctx, "Starting Gaze Indexer")
+					if err := indexer.Run(ctx); err != nil {
+						logger.PanicContext(ctx, "Something went wrong, error during running indexer", slogx.Error(err))
+					}
+				}()
+			}
+		}
 	}
 
+	// Run API server
+	httpServer := do.MustInvoke[*fiber.App](injector)
```

Comment:
There may or may not be an http server, depending on if we have a module with http handlers enabled or not.

Is it possible to start the http server only if such services exist instead?

Reply:
It's to complicated to detect and running http server without any enabled module has no effect for now.